File: stara-gradiva/hs-1-uvod/naloge/Uvod.hs
{-
 - Vaja 1: Uvod v Haskell
 -}

-- Funkcija predzadnjiElement vrne predzadnji element seznama l.
predzadnjiElement :: [a] -> a
predzadnjiElement l = last $ init l

-- Funkcija poisci naj poišče k-ti element v seznamu l.
-- Zgled:
-- ghci> poisci 2 [0,0,1,0,0,0]
-- 1
poisci :: Int -> [a] -> a
poisci k l = last $ take (k+1) l

-- Funkcija podvoji naj 'podvoji' seznam l.
-- Zgled:
-- ghci> podvoji [1,2,3,3]
-- [1,1,2,2,3,3,3,3]
-- Namig: Funkcija concat iz seznama seznamov naredil seznam z elementi podseznamov.
podvoji :: [a] -> [a]
podvoji l = concat [[x,x] | x <- l]

-- Funkcija razdeli k l naj seznam l razdeli na dva seznama. V prvem naj bo prvi k elementov
-- seznama l, v drugem pa vsi ostali. Funkcija naj vrne par teh dveh seznamov.
-- Zgled:
-- ghci> razdeli 2 [1,1,1,2,2,2]
-- ([1,1],[1,2,2,2])
razdeli :: Int -> [a] -> ([a], [a])
razdeli k l = (take k l, drop k l)

-- Funkcija zbrisi naj iz seznama l pobriše k-ti element.
-- Zgled:
-- ghci> zbrisi 3 [0,0,0,1,0,0,0]
-- [0,0,0,0,0,0]
zbrisi :: Int -> [a] -> [a]
zbrisi k l = take k l ++ drop (k+1) l

-- Funkcija rezina i k l naj sestavi novi seznam, ki naj vsebuje elemente seznama l od vključno
-- i-tega do -- k-tega (brez k-tega).
-- Zgled:
-- ghci> rezina 3 6 [0,0,0,1,2,3,0,0,0]
-- [1,2,3]
rezina :: Int -> Int -> [a] -> [a]
rezina i k l = drop i $ take k l

-- Napišite funkcijo "vstavi x k l", ki na k-to mesto seznama l vrine x.
-- Na primer:
-- ghci> vstavi 2 5 [0,0,0,0,0,0]
-- [0,0,0,0,0,2,0]
vstavi :: a -> Int -> [a] -> [a]
vstavi x k l = take k l ++ [x] ++ drop k l

-- Napišite funkcijo "zavrti n l", ki seznam l zavrti za n mest v levo.
-- Na primer:
-- ghci> zavrti 2 [1,2,3,4,5]
-- [3,4,5,1,2]
zavrti :: Int -> [a] -> [a]
zavrti n l = drop n l ++ take n l

-- Napišite funkcijo "pobrisi x l", ki vrne seznam, ki je kot l,
-- le da smo iz njega pobrisali vse pojavitve elementa x
-- Na primer:
-- ghci> pobrisi 'a' "abrakadabra"
-- "brkdbr"
pobrisi :: Eq a => a -> [a] -> [a]
pobrisi x l = [y | y <- l, y /= x]

-- Funkcija jePalindrom naj ugotovi, če je seznam l palindrom.
-- Zgled:
-- ghci> jePalindrom [1,2,3,2,1]
-- True
-- ghci> jePalindrom [1,2,2,1]
-- True
-- ghci> jePalindrom [1,2,3]
-- False
jePalindrom :: Eq a => [a] -> Bool
jePalindrom l =	l == reverse l

-- Napišite funkcijo "maxPoKomponentah l1 l2", ki vrne seznam, ki ima za
-- elemente večjega od elementov na ustreznih mestih v seznamih l1 in l2.
-- Če je eden od seznamov krajši, naj bo krajši tudi vrnjeni seznam.
-- Na primer:
-- ghci> maxPoKomponentah [1,10,5,6] [2,3,7,4,8]
-- [2,10,7,6]
maxPoKomponentah :: Ord a => [a] -> [a] -> [a]
maxPoKomponentah l1 l2 = [max x y | (x,y) <- zip l1 l2]

-- Napišite funkcijo "drugiNajvecji l", ki vrne drugi največji element
-- seznama l. Predpostavite lahko, da sta v l vsaj dva različna elementa.
-- Na primer:
-- ghci> drugiNajvecji [1,10,5,6]
-- 6
drugiNajvecji :: Ord a => [a] -> a
drugiNajvecji l = maximum $ pobrisi (maximum l) l

File: 4-uvod-v-ocaml/vaje/uvod.ml

(* ========== Vaja 1: Uvod v OCaml  ========== *)

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [predzadnji_element], ki vrne predzadnji element danega
 seznama. V primeru prekratkega seznama naj vrne napako.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # predzadnji_element [1; 2; 3; 4];;
 - : int = 3
[*----------------------------------------------------------------------------*)

let rec predzadnji_element = function
  | x :: _ :: [] -> x
  | _ :: xs -> predzadnji_element xs
  | [] -> failwith "List is too short."

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [poisci k lst], ki poišče [k]-ti element v seznamu [lst].
 Številčenje elementov seznama (kot ponavadi) pričnemo z 0. Če je k negativen,
 naj funkcija vrne ničti element.
 V primeru prekratkega seznama naj funkcija vrne napako.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # poisci 2 [0; 0; 1; 0; 0; 0];;
 - : int = 1
[*----------------------------------------------------------------------------*)

let rec poisci k = function
  | [] -> failwith "List is too short."
  | x :: xs -> if k <= 0 then x else poisci (k-1) xs

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [podvoji], ki podvoji pojavitve elementov v seznamu.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # podvoji [1; 2; 3];;
 - : int list = [1; 1; 2; 2; 3; 3]
[*----------------------------------------------------------------------------*)

let rec podvoji = function
  | x :: xs -> x :: x :: podvoji xs
  | [] -> []

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [razdeli k lst], ki seznam razdeli na dva seznama. Prvi
 naj vsebuje prvih [k] elementov, v drugem pa vsi ostali. Funkcija vrne par teh
 dveh seznamov.
 V primeru, ko je [k] izven mej seznama, naj bo primeren od seznamov prazen.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # razdeli 2 [1; 2; 3; 4; 5];;
 - : int list * int list = ([1; 2], [3; 4; 5])
 # razdeli 7 [1; 2; 3; 4; 5];;
 - : int list * int list = ([1; 2; 3; 4; 5], [])
[*----------------------------------------------------------------------------*)

let rec razdeli k lst =
  match (k, lst) with
  | (_, []) -> ([], [])
  | (k, lst) when k <= 0 -> ([], lst)
  | (k, x :: xs) ->
      let (lst1, lst2) = razdeli (k - 1) xs in
	    (x :: lst1, lst2)

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [zbrisi k lst], ki iz seznama izbriše [k]-ti element.
 V primeru prekratkega seznama naj funkcija vrne napako.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # zbrisi 3 [0; 0; 0; 1; 0; 0];;
 - : int list = [0; 0; 0; 0; 0]
[*----------------------------------------------------------------------------*)

let rec zbrisi k = function
  | [] -> failwith "List is too short."
  | x :: xs -> if k = 0 then xs else x :: zbrisi (k-1) xs

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [rezina i k lst], ki sestavi nov seznam, ki vsebuje
 elemente seznama [lst] od vključno [i]-tega do izključno [k]-tega.
 Predpostavimo, da sta [i] in [k] primerna.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # rezina 3 6 [0; 0; 0; 1; 2; 3; 0; 0];;
 - : int list = [1; 2; 3]
[*----------------------------------------------------------------------------*)

let rezina i k lst =
  let (_, rez1) = razdeli i lst in
  let (rez2, _) = razdeli (k - i) rez1 in
  rez2

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [vstavi x k lst], ki na [k]-to mesto seznama [lst] vrine
 element [x]. Če je [k] izven mej seznama, naj ga funkcija doda na začetek
 oziroma na konec.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # vstavi 1 3 [0; 0; 0; 0; 0];;
 - : int list = [0; 0; 0; 1; 0; 0]
 # vstavi 1 (-2) [0; 0; 0; 0; 0];;
 - : int list = [1; 0; 0; 0; 0; 0]
[*----------------------------------------------------------------------------*)

let rec vstavi x k = function
  | [] -> [x]
  | y :: ys -> if k <= 0 then x :: y :: ys else y :: vstavi x k ys

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [zavrti n lst], ki seznam zavrti za [n] mest v levo.
 Predpostavimo, da je [n] v mejah seznama.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # zavrti 2 [1; 2; 3; 4; 5];;
 - : int list = [3; 4; 5; 1; 2]
[*----------------------------------------------------------------------------*)

let rec zavrti n lst =
  let (zacetek, konec) = razdeli n lst in
  konec @ zacetek

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [pobrisi x lst], ki iz seznama [lst] izbriše vse pojavitve
 elementa [x].
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # pobrisi 1 [1; 1; 2; 3; 1; 2; 3; 1; 1];;
 - : int list = [2; 3; 2; 3]
[*----------------------------------------------------------------------------*)

let rec pobrisi x = function
  | y :: ys -> if y = x then pobrisi x ys else y :: pobrisi x ys
  | [] -> []

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [je_palindrom], ki za dani seznam ugotovi ali predstavlja
 palindrom.
 Namig: Pomagaj si s pomožno funkcijo, ki obrne vrstni red elementov seznama.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # je_palindrom [1; 2; 3; 2; 1];;
 - : bool = true
 # je_palindrom [0; 0; 1; 0];;
 - : bool = false
[*----------------------------------------------------------------------------*)

let je_palindrom lst =
  let rec obrni = function
    | x :: xs -> obrni xs @ [x]
	  | [] -> []
  in
  lst = obrni lst

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [max_po_komponentah], ki sprejme dva seznama in vrne nov
 seznam, katerega elementi so večji od istoležnih elementov na danih seznamih.
 Skupni seznam ima dolžino krajšega od danih seznamov.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # max_po_komponentah [5; 4; 3; 2; 1] [0; 1; 2; 3; 4; 5; 6];;
 - : int list = [5; 4; 3; 3; 4]
[*----------------------------------------------------------------------------*)

let rec max_po_komponentah lst1 lst2 =
  match (lst1, lst2) with
  | (x :: xs, y :: ys) -> max x y :: max_po_komponentah xs ys
  | _ -> []

(*----------------------------------------------------------------------------*]
 Definirajte funkcijo [drugi_najvecji], ki vrne drugo največjo vrednost v
 seznamu. Pri tem se ponovitve elementa štejejo kot ena vrednost.
 Predpostavimo, da ima seznam vsaj dva različna elementa.
 Namig: Pomagaj si s pomožno funkcijo, ki poišče največjo vrednost v seznamu.
 - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 # drugi_najvecji [1; 10; 11; 11; 5; 4; 10];;
 - : int = 10
[*----------------------------------------------------------------------------*)

let drugi_najvecji lst =
  let rec najvecji = function
    | [] -> failwith "List is too short."
	  | x :: [] -> x
	  | x :: xs -> max x (najvecji xs)
  in
  najvecji (pobrisi (najvecji lst) lst)
